### ISSUE TITLE
tabs: warn if internal link targets will be detected falsly

### ISSUE THREAD

**Issue by cagix** [McShelby/hugo-theme-relearn#681]
If you activate `externalLinkTarget: "_blank"` in the Hugo configuration, internal links (i.e. links to your site) within the `tabs`/`tab` shortcodes will be incorrectly recognised as external links and opened in a new tab.

**Comment by cagix** [comment]
Strangly this behaviour won't be observable when "hosting" from filesystem. 

This is (part of) my test code:

```markdown

{{< tabs groupid="pruefungsordnung" >}}
{{% tab title="Compilerbau" %}}

**Mündliche Prüfung plus Testat**, 5 ECTS (PO18)

-   **Testat**: Vergabe der Credit-Points
    ...

    Kriterien:

    -   Aktive Mitarbeit im [Praktikum]({{< ref "project" >}}) und an den Meilensteinen
    -   Aktive Teilnahme an den drei gemeinsamen Terminen mit der University of Alberta (Edmonton)
    -   ...

-   **Gesamtnote**:
    ...

{{% /tab %}}
{{% tab title="Concepts of Programming Languages" %}}

**Parcoursprüfung**, 10 ECTS (PO23)

...

{{% /tab %}}
{{< /tabs >}}


### Praktikum

Sie bearbeiten im Praktikum eine gemeinsame [Aufgabe]({{< ref "project" >}}) in der Semestergruppe.

```


The first reference leads to a link which will be opened in a new browser tab, the second reference to the same subpage will be opened in the same tab. It's in both cases a reference to an internal page `homework/project/_index.md`...

**Comment by McShelby** [comment]
"Works as designed".

If you use `ref` your link will be converted to a fully quallified link containing a leading `http://` or `https://`. This is the indicator that distinguish between external and internal. Internal links are missing the protocol part.

If you switch to `relref` it works as you intend.

**Comment by cagix** [comment]
OK? But why the different behaviour inside `tab`/`tabs` vs. "normal" text?

**Comment by McShelby** [comment]
That's really weird. For now you can replace your call with `[Aufgabe]({{% ref "project" %}})`

See: https://discourse.gohugo.io/t/render-link-hook-behaves-unexpected-for-different-shortcode-call-syntax/46601

**Comment by cagix** [comment]
Hmmm, on further investigation this seems not to be connected with `externalLinkTarget: "_blank"`. 

Whether this line is in my config or not, the following snippet

```markdown
[Praktikum REF]({{< ref "project" >}})

[Praktikum RELREF]({{< relref "project" >}})


{{< tabs groupid="pruefungsordnung" >}}
{{% tab title="Compilerbau" %}}

[Praktikum REF]({{< ref "project" >}})

[Praktikum RELREF]({{< relref "project" >}})

{{% /tab %}}
{{< /tabs >}}
```

will be translated to

```html
<p><a href="https://www.hsbi.de/elearning/data/FH-Bielefeld/lm_data/lm_1371719/homework/project.html">Praktikum REF</a></p>
<p><a href="https://www.hsbi.de/elearning/data/FH-Bielefeld/lm_data/lm_1371719/homework/project.html">Praktikum RELREF</a></p>

<div class="tab-panel" data-tab-group="pruefungsordnung">
  <div class="tab-nav">
    <div class="tab-nav-title">&#8203;</div>
    <button
      data-tab-item="compilerbau"
      class="tab-nav-button tab-panel-style cstyle initial active"
      onclick="switchTab('pruefungsordnung','compilerbau')"
    >
      <span class="tab-nav-text">Compilerbau</span>
    </button>
  </div>
  <div class="tab-content-container">
    <div
      data-tab-item="compilerbau"
      class="tab-content tab-panel-style cstyle initial active">
      <div class="tab-content-text">

<p><a href="https://www.hsbi.de/elearning/data/FH-Bielefeld/lm_data/lm_1371719/homework/project.html" target="_blank">Praktikum REF</a></p>
<p><a href="https://www.hsbi.de/elearning/data/FH-Bielefeld/lm_data/lm_1371719/homework/project.html">Praktikum RELREF</a></p>
</div>
    </div>
  </div>
</div>
```

I.e., inside the `tabs`/`tab` shortcode, `ref` and `relref` show different results (as explained in your last comment). However, outside this shortcode `ref` and `relref` do lead to the same link, which will open in the same browser tab. 

Feels like I'm missing something? Could this be related to me using ugly URLs?

**Comment by McShelby** [comment]
As linked above, I've opened an official request.

**Comment by cagix** [comment]
Yeah, sorry - posts crossing :)

Meanwhile, I've tried your suggestion:

1. `externalLinkTarget: "_blank"` **not** present: 
    - outside `tabs`/`tab`: 
        - `< ref >` => plain link
        - `< relref >` => plain link
        - `% ref %` => `target="_blank"`
        - `% relref %` => plain link
    - inside `tabs`/`tab`: 
        - `< ref >` => `target="_blank"`
        - `< relref >` => plain link
        - `% ref %` => `target="_blank"`
        - `% relref %` => plain link
2. `externalLinkTarget: "_blank"` present: 
    - outside `tabs`/`tab`: 
        - `< ref >` => plain link
        - `< relref >` => plain link
        - `% ref %` => `target="_blank"`
        - `% relref %` => plain link
    - inside `tabs`/`tab`: 
        - `< ref >` => `target="_blank"`
        - `< relref >` => plain link
        - `% ref %` => `target="_blank"`
        - `% relref %` => plain link

So it seems, that in my case `externalLinkTarget: "_blank"` seems to be not relevant to the result. (Plus the difference between the `<...>` and the `%...%` form inside vs. outside.)

**Comment by McShelby** [comment]
It's a damn rabbit hole.

The official recommendation is to use plain Markdown links.

Otherwise it is mandatory calling `ref` / `relref` with `%` in case of this theme (and all other themes that might implement similar functionality)

I'll now take a further look into your posted examples.

**Comment by cagix** [comment]
> It's a damn rabbit hole.

it definitely feels like it, i've been reading your discourse discussion. 


> 
> The official recommendation is to use plain Markdown links.

that could turn out to be a bit difficult in my use case 🙈



> Otherwise it is mandatory calling `ref` / `relref` with `%` in case of this theme (and all other themes that might implement similar functionality)

could live with that 👍


>
> I'll now take a further look into your posted examples.

thanks a lot for your work and quick support!

**Comment by McShelby** [comment]
So let's disect this.

- if `ref` and `relref` are used, you need to call it with `%` (so all cases using `<` are off the table as they will never work as expected)
- `relref` will always be treated as internal link and _should_ open the link in the current browser tab (so all cases using `relref` are off the table as they behave as expected)
- `ref` will always be treated as external link and _should_ respect the `externalLinkTarget`
- not setting `externalLinkTarget` in your `config.toml` is equal to `externalLinkTarget="_blank"`
- `target=_blank` will [open a new browser/tab](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#target)

So, if you set `externalLinkTarget="_blank"` or don't setting it at all, your assumption

> The first reference leads to a link which will be opened in a new browser tab (unexpected), the second reference to the same subpage will be opened in the same tab (expected). 

is twisted. It's the other way around. The first is expected behavior, the second is (somewhat) not.

In respect of all the limitations and assumptions above, for me it seems everything works due the constraints. But correct if I am wrong.

Btw: If you want to open external links in the current tab, you can set `externalLinkTarget="_self"`.

**Comment by cagix** [comment]
Oh, I'm really sorry about that. I failed to realise that `externalLinkTarget = "_blank"` is the _default_, so whether it's present or not, it should always produce the same result (which it does). (I noticed the unexpected behaviour related to the `ref` only after I had "activated" this option and so I associated former with the latter.)

So, the take away for me is that I have to use the `%` format instead of the `<` format because the way Hugo handles things internally. Additionally, I want to use `relref` instead of `ref`, because I want to generate an internal link with the `ref`.

My initial issue is solved with this. However, the `%` vs. `<` could be something to be investigated or to be documented regarding the theme, so I'll leave the ticket open. Thank you very much again for your support and all the work on this great theme!
